### DISCUSSION TITLE
WIP: Tower of the Valley

### DISCUSSION THREAD

**Discussion by DaVince** [excaliburjs/Excalibur#1895] [Show and tell]
Are development threads for a game okay too? :smile: 

I've been away for a little while, but got motivated to use Excalibur again for a game I'm working on. It's called Tower of the Valley and so far it looks like this (don't pay too much attention to the programmer art!):

![118010028-ebcf8980-b34e-11eb-8155-eb73c92efbaf-2](https://user-images.githubusercontent.com/38518308/125014592-9e605780-e033-11eb-90e4-2600261f9052.png)

If this seems in any way familiar, it's because I'm basing the basic look and feel, as well as the game mechanics, on an old Windows 3.1 game called [Castle of the Winds](https://classicreload.com/castle-or-the-winds.html)! I'm intending to give it a sort of rogue-lite twist, though, where playthroughs have a limited length and replaying the game will unlock new things. In that way it'll kinda be like Kingsway, I suppose.

So this thing is still very much in early development but I wanted to at least share how things are going from time to time. :slightly_smiling_face:

**Comment by DaVince** [comment]
So far, I've implemented:

- Basic tile-based, eight-directional movement.
- An input mapper, responding to several keys that are defined to perform the same 'action', as well as gamepad input (presumably; I haven't actually tested the code for that yet).
- Map loading and switching.
- Tiled maps integration: I use the object layer cleverly to define things like the player starting position, areas that warp you to other areas, definitions of walls that have hidden doors, text to display, and soon, events to trigger.
- Normal, hidden and locked doors. The door tile gets replaced with an opened one when going through, too.
- An adventure log that keeps track of your actions.
- A randomly generated, potty-mouthed "You swing your ... ... sword" text. (Just a test/for personal entertainment)

Sorry if this "feature list" is a little technical, I just like to keep track of what systems I've managed to implement so far.

Next up I'm looking into a fog mechanic where the map isn't visible until you start navigating through it, just like in the original. But I'm having some difficulties with the implementation.

**Reply by DaVince** [reply]
Added:
* Search/wait mechanic (uncovers hidden doors)
* Enemies that move towards you
* Attacks, damage calculation, critical hits, enemy death
* Highlight colors in the adventure log

And refactored things here and there.

![image](https://user-images.githubusercontent.com/105533/119640883-57b6f500-be19-11eb-8718-3acc4bfd34c0.png)

(Updated to remove words some might find offensive; the flavor words on the sword have a mildly cussy, lighthearted and untargeted intention but I understand if it goes against the code of conduct here.)

**Reply by SimponGoril** [reply]
Man, I love Castle of the Winds. Thank you for bringing it back :)

**Reply by DaVince** [reply]
> Man, I love Castle of the Winds. Thank you for bringing it back :)

Thank you! Truth be told, I stopped developing this some time ago, but your words are encouraging me to start over. Excalibur has improved since then too so it's a good moment to try again (and also aim a little _smaller_, heh).

**Comment by DaVince** [comment]
I restarted my code base based on the latest version of Excalibur and the game looks a fair bit different now. Created a GUI system with a faux 90's OS look to it. I've also lost a fair amount of mechanics that I'm going to reimplement because the previous code base was a mess.

![2022-07-15 tower-of-the-valley](https://user-images.githubusercontent.com/105533/179216001-6a58a4b7-7572-40c0-9ab6-224975261f8f.png)

**Reply by eonarheim** [reply]
@DaVince Very cool! I'd love your take on building a GUI system in Excalibur, that looks really good!

**Reply by DaVince** [reply]
Well, I have to be honest... My brain still doesn't seem to want to get the hang of ECS so it's a bunch of classes and subclasses that make use of extending Actors. Windows use child actors for the elements they're composed of, and windows/buttons/panels of arbitrary sizes use GraphicsGroups. My implementation is hacky but functional; probably not something too suitable to share as a common library. :sweat_smile:

Still, the code is going to be open source once I feel confident enough that it's complete enough to even share, so I'll be sure to share it once that time comes.
